cpp/src/Ice/Reference.cpp

```diff
@@ -1549,136 +1549,74 @@ IceInternal::RoutableReference::createConnectionAsync(
-            {
-                _exception(ex);
-            }
-
-            CB1(
-                const RouterInfoPtr& routerInfo,
-                function<void(ConnectionIPtr, bool)> response,
-                function<void(std::exception_ptr)> exception) :
-                _routerInfo(routerInfo), _response(std::move(response)), _exception(std::move(exception))
-            {
-            }
-
-        private:
-
-            const RouterInfoPtr _routerInfo;
-            const function<void(ConnectionIPtr, bool)> _response;
-            const function<void(std::exception_ptr)> _exception;
-        };
+            connection->setAdapter(self->getRouterInfo()->getAdapter());
+        }
+        response(connection, compress);
+    };
 
-        //
-        // Get an existing connection or create one if there's no
-        // existing connection to one of the given endpoints.
-        //
-        factory->create(endpoints, false, getEndpointSelection(), make_shared<CB1>(_routerInfo, response, exception));
+    if(getCacheConnection() || endpoints.size() == 1)
+    {
+        // Get an existing connection or create one if there's no existing connection to one of the given endpoints.
+        factory->createAsync(
+            endpoints,
+            false,
+            getEndpointSelection(),
+            std::move(createConnectionResponse),
+            std::move(exception));
         return;
     }
     else
     {
-        class CB2 final :
-            public OutgoingConnectionFactory::CreateConnectionCallback,
-            public std::enable_shared_from_this<CB2>
+        // Go through the list of endpoints and try to create the connection until it succeeds. This is different from
+        // just calling create() with the given endpoints since this might create a new connection even if there's an
+        // existing connection for one of the endpoints.
+
+        class CreateConnectionState final
         {
         public:
+            exception_ptr exception = nullptr;
+            size_t endpointIndex = 0;
+            function<void(exception_ptr)> createConnectionException;
+        };
 
-            void setConnection(const Ice::ConnectionIPtr& connection, bool compress) final
-            {
-                //
-                // If we have a router, set the object adapter for this router
-                // (if any) to the new connection, so that callbacks from the
-                // router can be received over this new connection.
-                //
-                if(_reference->getRouterInfo() && _reference->getRouterInfo()->getAdapter())
-                {
-                    connection->setAdapter(_reference->getRouterInfo()->getAdapter());
-                }
-                _response(connection, compress);
-            }
-
-            void setException(std::exception_ptr ex) final
+        auto state = make_shared<CreateConnectionState>();
+        state->createConnectionException =
```

Comment:
Is this code leak-free?

My concern:
- CreateConnectionState holds onto std::function object (createConnetionException)
- std::function object own a refcount on shared_ptr<CreateConnectionState> (in lambda capture)

Reply:
Updated the code to clear `CreateConnectionState`